magni_bringup/scripts/launch_core.py

```diff
@@ -68,10 +70,26 @@ def get_conf():
         print "Error calling pifi"
         subprocess.call(["chronyc", "waitsync", "6"]) # Wait up to 60 seconds for chrony
 
+    boardRev = 0
+
+    if conf['motor_controller']['board_version'] == None: 
+        # Code to read board version from I2C
+        # The I2C chip is only present on 5.0 and newer boards
+        try:
+            i2cbus = smbus.SMBus(1)
+            i2cbus.write_byte(0x20, 0xFF)
+            time.sleep(0.2)
+            inputPortBits = i2cbus.read_byte(0x20)
+            boardRev = 49 + (15 - (inputPortBits & 0x0F))
+            print "Got board rev: %d" % boardRev
+        except: 
+            print "Error reading board version from i2c"
+
     # Ugly, but works 
     # just passing argv doesn't work with launch arguments, so we assign sys.argv
     sys.argv = ["roslaunch", "magni_bringup", "core.launch", 
                          "raspicam_mount:=%s" % conf['raspicam']['position'],
-                         "sonars_installed:=%s" % conf['sonars']]
+                         "sonars_installed:=%s" % conf['sonars'],
+                         "controller_board_version:=%d" % boardRev]
```

Comment:
Here is where it gets tricky.  My thinking is IF we can read I2C AND get a firmware bitfield that is between 1 and 14 (in other words, NOT all zero or all 1) we 'trust' that value and use it for controller board version just like your logic.    The tricky logic is if we cannot read a valid ID from I2C we only pass in controller_board_version if the user has NOT put his own controller_board_version in the yaml file.   My thoughts are it is always nice if a user can force something and when a user forces something we believe the human and do NOT try to correct him/her unless we are really sure we were unable to read from I2C a version that makes sense.

Reply:
I agree with choice 1. If a user manually sets a version, they must have a deliberate and important purpose for it, so we should respect their choice. For example if they are on a non-raspberry pi platform and they don't want to write anything to that I2C port.

Reply:
Ok, Agreed. Choice 1.  If they set it, we respect it, whatever it may be.